File: trunk/PSC/SERIES/Serie05/Docs/PSC_Y0910T2_LI31N_SE5-1_Attachment_FILES/Board.java

public final class Board {
	public static final int LINES=6, COLS=6, BOMBS=5;

	private final Cell[][] cells = new Cell[LINES][COLS];
	
	private int bombs; // number of bombs not flagged. 
	private int hides; // number of cells not touched.

	public boolean isSolved() { return hides==bombs; }

	public Board() {
	  hides = LINES*COLS;
	  for( ; bombs < BOMBS ; ++bombs)
		putBomb((int)(Math.random()*(hides-bombs)));
	  for(int l=0 ; l<LINES ; ++l)
	    for(int c=0; c<COLS ; ++c)
	   	  if (cells[l][c]==null)
	   		cells[l][c]=new EmptyCell(l,c);
	}
	
	private void putBomb(int n) {
	  assert n>=0 : n;
	  for(int l=0 ; l<LINES ; ++l)
		for(int c=0; c<COLS ; ++c)
		  if (cells[l][c]==null) {
			  if (n==0) { cells[l][c]=new BombCell(); return; }
	          --n;
		  }
	  assert false : n;
	}
	
	private static void printLine() {
		  System.out.print("   +");
		  for(int c=0; c<COLS ; ++c) System.out.print("--");
		  System.out.println("-+");
	}

	public void print() {
	  System.out.printf("%02d  ",bombs);
	  for(int c=0; c<COLS ; ++c) System.out.printf(" %c", 'A'+c);
	  System.out.printf("   %02d", hides);
	  System.out.println();
	  printLine();
	  for(int l=0 ; l<LINES ; ++l) {
		System.out.printf("%2d |",l+1);
		for(int c=0; c<COLS ; ++c) {
			System.out.print(' ');
			cells[l][c].print();
		}
		System.out.println(" |");
	  }
	  printLine();
	}
	
	public static boolean isValid(int l, int c) {
      return l >= 0 && l < Board.LINES && c >= 0 && c < Board.COLS;	
	}

	public boolean isBomb(int l, int c) {
	  return isValid(l,c) && cells[l][c].isBomb();
    }

	public void touch(int l, int c) {
	  if (isValid(l, c)) {
		  Cell cell = cells[l][c];
		  if (cell.isShown() || cell.isFlagged()) return;
		  cell.touch();
		  hides--;
	  }
	}

	public void showAll() {
	  for(int l=0 ; l<LINES ; ++l)
		for(int c=0; c<COLS ; ++c) cells[l][c].show();
	  hides = 0;
	}

	public void flag(int l, int c) {
		if (isValid(l,c)) {
			  Cell cell = cells[l][c];
			  if (cell.isShown()) return;
			  cell.toggleFlag();
			  if (cell.isFlagged()) { bombs--; hides--; }
			  else { bombs++; hides++; }
		}
	}
}

File: trunk/PSC/SERIES/Serie05/Src/Board.c
#include "Board.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <math.h>
#include <time.h>
#include <string.h>
#include "BombCell.h"
#include "EmptyCell.h"
/**
 * Método que coloca a bomba no tabuleiro
 * */
static void Board_putBomb(Board* this,int n){
	int l,c;
	assert(n>=0);
	for(l=0 ; l<LINES ; ++l){
		for(c=0; c<COLS ; ++c){
			if (this->cells[l][c] == NULL) {
				if (n==0) { 
				  this->cells[l][c]=(Cell*)BombCell_new();
				  return; 
				}
			--n;
			}
		 }
	}
	  assert(n > 0);	
}
/**
 * Método que imprime uma linha do border do tabuleiro
 * */
static void Board_printLine(Board* this){
		int c;
		printf("   +");
		for(c =0;c < COLS;++c) printf("--");
		puts("-+");
}

/**
 * Métodos Booleanos que verificam se a posição é válida, o tabuleiro está resolvido,
 * se o jogo já acabou ou se uma posição é uma bomba.
 * */
boolean Board_isValid(Board* this, int l, int c){return ( l >= 0 && l < LINES && c >=0 &&  c < COLS)?true:false;}
boolean Board_isSolved(Board* this){return this->hides == this->bombs?true:false;}
boolean Board_isOver(Board* this){return this->over;}
boolean Board_isBomb(Board* this, int l, int c){
	return (Board_isValid(this, l, c) && this->cells[l][c]->vptr->isBomb(this->cells[l][c]))?true:false;
}

/**
 * Imprimem o tabuleiro no ecra
 * */
void Board_print(Board* this){
	int c,l;
	printf("%02d  ",this->bombs);
	for(c=0; c < COLS ; ++c) printf(" %c", 'A'+c);
	printf("   %02d\n", this->hides);
	Board_printLine(this);
	for(l=0 ; l<LINES ; ++l) {
		printf("%2d |",l+1);
		for(c=0; c<COLS ; ++c) {
			printf(" ");
			Cell_print(this->cells[l][c]);
		}
		puts(" |");
	}
	Board_printLine(this);
}
/**
 * Procede à modificação de uma posição
 * */
void Board_touch(Board* this, int l, int c){
	if (Board_isValid(this, l, c)){
		Cell* cel = this->cells[l][c];
		if (Cell_isShown(cel) || Cell_isFlagged(cel)){ return;}
		cel->vptr->touch(cel,this);
		--(this->hides);
	}
}
/**
 * Torna todas as posições visiveis
 * */
void Board_showAll(Board* this){
	int l,c;
	this->over = true;
	for (l=0;l< LINES;++l){
		for(c=0;c< COLS;++c){
			Cell_show(this->cells[l][c]);
		}
	}
	
}
/**
 * Procede à modificação de uma posição, marcando-a como flag
 * */
void Board_flag(Board* this, int l, int c){
	if (Board_isValid(this,l,c)){
		Cell* cel = this->cells[l][c];
		if (Cell_isShown(cel)) return;
		Cell_toggleFlag(cel);
		if (Cell_isFlagged(cel)){(this->bombs)--;(this->hides)--;}
		else{(this->bombs)++;(this->hides)++;}
	}
}

/*
 * (Des)Construtores e (Des)Iniciadores
 * */
static void Board_delete_array(Board* this){
	int l,c;
	for(l=0 ; l<LINES ; ++l){
		for(c=0; c<COLS ; ++c)
			this->cells[l][c]->vptr->dtor(this->cells[l][c]);
	
	}
}
void Board_Cleanup(Board* this){
	if (this == NULL){ 
		return;
	}else{
		Board_delete_array(this);
	}
}
void Board_delete(Board* this){
	if (this == NULL) return;
	Board_Cleanup(this);
	free(this);
}
static void Board_init_array(Board* this){
	int l,c;
	for(l=0 ; l<LINES ; ++l){
		for(c=0; c<COLS ; ++c)
			this->cells[l][c]=NULL;
	
	}
}

void Board_init(Board* this){
	int l,c;
	unsigned int r;
	if (this == NULL) return;
	
	this->hides=LINES*COLS;
	this->bombs=0;
	this->over=false;
	
	Board_init_array(this);
	for( ; this->bombs < BOMBS ; ++this->bombs){
		r=(unsigned int)rand();	
		Board_putBomb(this,r%(this->hides - this->bombs));
		srand(time(NULL));
	}
	for(l=0 ; l<LINES ; ++l){
		for(c=0; c<COLS ; ++c)
			if ((this->cells[l][c])==NULL){
				this->cells[l][c]=(Cell*)EmptyCell_new(l,c);
			}
	}

}

Board* Board_new(){
	Board* this=(Board*)calloc(1,sizeof(Board));
	assert(this == NULL);
	Board_init(this);
	return this;
}

/**
 * Copia o conteudo de um tabuleiro para outro.
 * Util para fazer load de um jogo
 * */
static void	Board_Copy(Board* dest, Board* src){
	int l,c;
	if(dest == NULL || src == NULL) return;
	Board_Cleanup(dest);
	Board_init_array(dest);
	dest->hides = src->hides;
	dest->bombs = src->bombs;
	for(l=0;l<LINES;++l){
		for(c=0;c<COLS;++c){
			dest->cells[l][c]=src->cells[l][c];
			dest->cells[l][c]->stat=src->cells[l][c]->stat;
		}
	}
}
/**
 * Exporta um tabuleiro no formato
 * <TYPE><STAT>
 * Aloca dinamicamente espaço para uma string com o tabuleiro.
 * quem chama a função deve libertar o espaço quando não necessitar 
 * */
char* Board_export(Board* this){
	if (this == NULL) return NULL;
	
	if (Board_isSolved(this) || Board_isOver(this)) {puts("No Game to Save!"); return NULL;}
	else{
	int i, j;
	char* return_str=(char*)calloc(LINES*COLS*2+1,sizeof(char));
	char* ret=return_str;
	
	for(i=0;i<LINES;++i){
		for(j=0;j<COLS;++j,return_str+=2){
			sprintf(return_str,"%c%c",this->cells[i][j]->type,this->cells[i][j]->stat);
		}
	}
			
	*(return_str ) = 0;
	
	return ret;
	}
}
/**
 * Contabiliza as bombas para o carregamento de um jogo gravado
 * */
static void Board_import_count(Board* this){
	int l,c;
	Cell* cel;
	for (l=0;l<LINES;++l){
		for(c=0;c<COLS;++c){
			cel=this->cells[l][c];
			if (cel->type == 'E' && Cell_isShown(cel)){
				EmptyCell_touch_count((EmptyCell*)this->cells[l][c],this);
			}
		}
	}
}
/**
 * Importa um jogo Gravado
 * */
void Board_import(Board* this, char* str){
	int cellnbr=0;
	Board savedBoard;
	if (this ==  NULL && str == NULL) {puts("No Game to Load!");return;}
	
	Board_init_array(&savedBoard);
	savedBoard.hides=LINES*COLS;
	savedBoard.bombs=0;
	savedBoard.over=false;
	for(;*str;++str,++cellnbr){
		if (*str == 'B'){
			savedBoard.cells[cellnbr/LINES][cellnbr%COLS]=(Cell*)BombCell_new();
			savedBoard.bombs++;
		}else if (*str == 'E'){
			savedBoard.cells[cellnbr/LINES][cellnbr%COLS]=(Cell*)EmptyCell_new(cellnbr/LINES,cellnbr%COLS);	
			
		}else{
			puts("Invalid Board load.");
			Board_Cleanup(&savedBoard);
			return;
		}
		str++;	
		savedBoard.cells[cellnbr/LINES][cellnbr%COLS]->stat=*str;
	}
	Board_Copy(this,&savedBoard);
	
	Board_import_count(&savedBoard);
	
}
